vision data and supported tasks and formats (#2291)
deepchecks/deepchecks: docs/source/user-guide/vision/VisionData.rst

.. _vision_data_class:

========================
The Vision Data Class
========================
The :class:`VisionData <deepchecks.vision>` data class is the base deepchecks class for
representing the data for a vision task. It is essentially a wrapper around a batch loader of images, labels,
and predictions in :doc:`Deepchecks' format </user-guide/vision/supported_tasks_and_formats>`
which is used as the input for the different :doc:`vision checks </checks_gallery/vision>`.

> In addition it stores cache regarding the data, labels and predictions it encountered during the check run.

Reviewer: ```suggestion
In addition it stores cached information about the data, labels and predictions it encountered during the check run.
```

Class Properties
================

The common properties are:

- **batch_loader** - The batch loader can be either a pytorch DataLoader, a tensorflow Dataset or any custom

>   generator. It is important to note that the data loaded by the

Reviewer: why use the word generator and not iterator?

Author: the return type of the function will be - t.Generator

  batch loader iterator must be **shuffled**.

- **task_type** - The task type of the data, can be either ``classification``, ``object_detection``,
  ``semantic_segmentation``, or ``other``. Data format validation is done upon creation of VisionData based
  on the selected task type.

- **label_map** - A dictionary mapping class ids to their names.

Creating a VisionData Object
============================

> 

Reviewer: Maybe should mention that these are simple examples without predictions, and that for predictions there will be an additional section (after you finish reading this)

From Generic Generator
----------------------

> If you are not already using pytorch or tensorflow for the project, this is the most recommended option.

Reviewer: Maybe mention explicitly other python frameworks like fastai, jax, non-python frameworks, autoML models etc...

The custom generator can be implemented in any way you like, as long as it outputs the data in the
:doc:`required format </user-guide/vision/supported_tasks_and_formats>`

> and that it loads the data in a shuffled manner. The following is an example of a custom generator based on data that

Reviewer: Also a batch should fit in memory.

Author: it says all the data is stored in memory here

is loaded and stored in memory as numpy arrays.

> For an example of a custom generator that loads the data from disk batch by batch see the

Reviewer: In the paragraph above you're saying "The following example is a custom generator loading from disk", and here you're saying "to see custom generator from disk go somewhere else" which confused me.

:ref:`following section <vision_data__precalculated_predictions>`.

.. code-block:: python

    from deepchecks.vision import VisionData, BatchOutputFormat

    def custom_generator(batch_size = 64):
        images, labels = load_data(shuffle=True)
        for i in range(0, len(images), batch_size):
            # Extracting images and label for batch and converting images of (N, C, H, W) into (N, H, W, C)
            images_in_batch = images[i:i+batch_size].transpose((0, 2, 3, 1))
            labels_in_batch = labels[i:i+batch_size]
            # Convert ImageNet format images into to [0, 255] range format images.
            mean, std  = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
            images_in_batch = np.clip(std * images_in_batch + mean, 0, 1) * 255
            yield BatchOutputFormat(images= images_in_batch, labels= labels_in_batch)

>     vision_data = VisionData(custom_generator(), task_type='classification', shuffle_batch_loader=False)

Reviewer: After saying that data must be shuffled independently, not clear what shuffle_batch_loader does.

Author: would you add a comment above or remove the variable ? (user will get a warning when running the code)

>     vision_data.head()

Reviewer: What is that? Maybe a comment

From Pytorch DataLoader
-----------------------

> In order to create a pytorch DataLoader, you must first create a pytorch Dataset object (link). When creating a pytorch

Reviewer: forgot (link)

> In order to create a pytorch DataLoader, you must first create a pytorch Dataset object (link). When creating a pytorch

Reviewer: Maybe start by saying "you know DataLoaders? You can use you're own by just modifying the collate function!". Also not clear why we need to explain about the structure of a pytorch Dataset.

Author: updated

Dataset object, you must implement either the ``__get_item__`` function or the ``__iter__`` function, whose output is
a single data element.

Based on the output format we will create a collate function that will group several data
elements into a batch in :doc:`Deepchecks' format </user-guide/vision/supported_tasks_and_formats>`.
This collate function will be used as the ``collate_fn`` argument in the creation of the DataLoader.

.. code-block:: python

    import torch
    from torch.utils.data import DataLoader
    from deepchecks.vision import VisionData, BatchOutputFormat

    def deepchecks_collate(data) -> BatchOutputFormat:
        # Extracting images and label and converting images of (N, C, H, W) into (N, H, W, C)
        images = torch.stack([x[0] for x in data]).permute(0, 2, 3, 1)
        labels = [x[1] for x in data]
        # Convert ImageNet format images into to [0, 255] range format images.
        mean, std  = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
        images = np.clip(std * images.numpy() + mean, 0, 1) * 255
        return BatchOutputFormat(images= images, labels= labels)

>     data_loader = DataLoader(my_dataset, batch_size=64, generator=torch.Generator(),

Reviewer: remove `torch.Generator()`, not important here

                             collate_fn=deepchecks_collate)
    vision_data = VisionData(data_loader, task_type='classification')
    vision_data.head()

From TensorFlow Dataset
-----------------------
There are two possible ways to create a deepchecks compatible tensorflow
`Dataset object <https://www.tensorflow.org/api_docs/python/tf/data/Dataset>`_. You can either create it

> in a way that directly outputs the data in the required format or convert an existing dataset,

Reviewer: ```suggestion
in a way that directly outputs the data in the required format or convert an existing dataset.
```

> in a way that directly outputs the data in the required format or convert an existing dataset,

Reviewer 2: ```suggestion
so that it outputs the data in the required format, or convert an existing dataset,
```

> we will demonstrate the latter.

Reviewer: ```suggestion
We will demonstrate the latter option.
```

> Assume a tensorflow dataset object that outputs a batch of images and labels as a tuple of (images, labels).

Reviewer 2: ```suggestion
If we already have a dataset object that outputs a batch of images and labels as a tuple of (images, labels),
```
LOL this is not a math exercise 😆

> We will use the `map <https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map>`_

Reviewer 2: ```suggestion
we can use the `map <https://www.tensorflow.org/api_docs/python/tf/data/Dataset#map>`_
```
(continues the previous sentence)

function to convert the data into :doc:`Deepchecks' format </user-guide/vision/supported_tasks_and_formats>`.

.. code-block:: python

    from deepchecks.vision import VisionData, BatchOutputFormat

    def deepchecks_map(batch) -> BatchOutputFormat:
        # Extracting images and label and converting images of (N, C, H, W) into (N, H, W, C)
        images = batch[0].permute(0, 2, 3, 1)
        labels = batch[1]
        # Convert ImageNet format images into to [0, 255] range format images.
        mean, std  = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
        images = np.clip(std * images.numpy() + mean, 0, 1) * 255

>         return BatchOutputFormat(images= images, labels= labels)

Reviewer 2: space after "=". Also, do we prefer to show examples with dictionary or class?

Reviewer 2: Don't care enough :) Though there's still space after the "="

    deepchecks_dataset = my_dataset.map(deepchecks_map)
    vision_data = VisionData(deepchecks_dataset, task_type='classification')
    vision_data.head()

> 

Reviewer 2: Too many spaces

Adding Model Predictions
========================

> Some of deepchecks tests, including the :doc:`model evaluation checks and suite </checks_gallery/vision>`,

Reviewer 2: ```suggestion
Some checks, including the :doc:`model evaluation checks and suite </checks_gallery/vision>`,
```

require model predictions in

> order to run. Models prediction are supplied via the batch loader in a similar fashion to the images and labels.

Reviewer 2: ```suggestion
order to run. Model predictions are given via the batch loader in a similar fashion to the images and labels.
```

> There are several ways to supply them which can be divide into two categories: Pre-calculated predictions and

Reviewer 2: ```suggestion
There are 2 ways to input model predictions: Pre-calculate predictions, and
```
I know these can be split into additional ways, but no need to complicate this at this point IMO

on-demand inference.

> On-demand Inference

Reviewer 2: Do the sections in the order you mentioned them in the previous sentence

-------------------

> In this case we will need to incorporate the model object in the relevant format transformation function. This can be

Reviewer 2: ```suggestion
In this case we will need to incorporate the model object in our batch loader. For an existing batch loader, this means that the model object will be part of our collate function (pytorch) or map function (tensorflow). This can be
```

> done either by using the model as a global variable, creating a wrapper class for the transformation function or

Reviewer 2: ```suggestion
done either by using the model as a global variable, creating a wrapper class for the collate/map function or
```

Reviewer 2: I think that "transformation function" is a scary phrase so suggested just to name the common ones.

> creating a closure function. We will demonstrate the last option via the pytorch interface.

Reviewer: What should I do if I have Tensorflow? Think it's worth mentioning, at least in words.

Author: added

.. code-block:: python

    import torch
    from torch.utils.data import DataLoader
    from deepchecks.vision import VisionData, BatchOutputFormat

>     def create_deepchecks_collate(model):

Reviewer: Add device

        def deepchecks_collate(data) -> BatchOutputFormat:
            # Extracting images and label and predicting using the model
            raw_images = torch.stack([x[0] for x in data])
            predictions = model(images)
            labels = [x[1] for x in data]
            # Convert ImageNet format images into to [0, 255] range format images.
            mean, std  = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
            images = np.clip(std * raw_images.permute(0, 2, 3, 1).numpy() + mean, 0, 1) * 255
            return BatchOutputFormat(images= images, labels= labels, predictions= predictions)
        return deepchecks_collate

>     data_loader = DataLoader(my_dataset, batch_size=64, generator=torch.Generator(),

Reviewer: remove generator

                             collate_fn=create_deepchecks_collate(my_model))
    vision_data = VisionData(data_loader, task_type='classification')
    vision_data.head()

.. _vision_data__precalculated_predictions:
Pre-calculated Predictions
--------------------------

> It is specifically recommended to use this option if your model object is unavailable locally (for example if

Reviewer 2: ```suggestion
It is recommended to use this option if your model object is unavailable locally (for example if
```
Would actually rephrase further - 
"This is the recommended option, as pre-calculating the predictions saves time during the checks' run, and is more efficient as sometimes you would need to rerun your checks to see if you were able to fix your data issues."

Or something like that. And put this option first.

Why? Because if the user does this outside of the check, they would think our checks are faster

> placed on a separate prediction server) or if the predicting process is computationally expensive or time consuming.

Reviewer 2: in this context, "computationally expensive or time consuming" kind of mean the same thing.

In the example below we will read the pre-calculated predictions, as well as the images and labels, from
a csv file containing the path to the image, the label and the prediction probabilities per sample.

.. code-block:: python

    from PIL import Image
    from deepchecks.vision import VisionData, BatchOutputFormat

    def data_from_file_generator(batch_size = 64):
        # Shuffling is a must for generic generators in order to achieve accurate results

>         data = pd.read_csv('classification_data.csv', index_col=0).sample(frac=1)

Reviewer 2: Wondering if regular vision users would understand that the small ".sample(frac=1)" is the shuffling. Maybe we should consider doing this in a separate row?

        for i in range(0, len(data), batch_size):
            images = [Image.open(x) for x in data['path_to_image'][i:(i + batch_size):]]
            labels = data['label'][i:(i + batch_size):]
            prediction_probabilities_as_str = data['prediction_probabilities'][i:(i + batch_size):]

>             prediction_probabilities_as_arr = [x.strip('][').split(', ') for x in prediction_probabilities_as_str]

Reviewer: maybe simpler to write an example from json file?

Author: I dont think it matters enough to rewrite as we state explicitly in can be done in many different ways.
can to an example from json in the tutorial
